### PR TITLE
Support multi-modal inputs for PGD and FGM attacks

### LINKED ISSUE #623: Support for attack budget eps per feature
Many datasets like multi-modal or tabular data would benefit from an attack budget `eps` per mode or column to avoid having to normalise the data into a common data range which is not always possible e.g. for pretrained models.

Therefore this issue should explore and implement changes to support the attack budgets `eps of evasion attacks`, which so far are a single `float`, as possibly a broadcastable array that allows defining `eps` per feature.

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
Support for attack budget eps per feature to enable attacks on multi-modal input data by making `eps` broadcastable onto input `x`.

Fixes #623 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/fast_gradient.py [hunk 1]
```diff
@@ -102,6 +102,22 @@ def __init__(
         self._project = True
         FastGradientMethod._check_params(self)
 
+    def _check_eps(self, x: np.ndarray):
+        """
+        Check the correctness of the `eps` attribute, and implicitly check the correctness of the `eps_step`
+        attribute because `eps` and `eps_step` have same shape.
+
+        :param x: An array with the original inputs.
+        """
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:111]:
```suggestion
    def _check_compatibility_input_and_eps(self, x: np.ndarray):
        """
        Check the compatibility of the input with `eps` and `eps_step` which are of the same shape.

        :param x: An array with the original inputs.
        """
```

File: art/attacks/evasion/fast_gradient.py [hunk 2]
```diff
@@ -167,6 +236,9 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
         :type mask: `np.ndarray`
         :return: An array holding the adversarial examples.
         """
+        # Ensure eps is broadcastable
+        self._check_eps(x=x)
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:240]:
```suggestion
        self._check_compatibility_input_and_eps(x=x)
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 3]
```diff
@@ -191,11 +191,37 @@ def _check_params(self) -> None:
...
+            if (self.eps_step <= 0).any():
+                raise ValueError("The perturbation step-size `eps_step` has to be positive.")
+
+            if self.eps.shape != self.eps_step.shape:
+                raise ValueError(
+                    "The perturbation size `eps` and the perturbation step-size `eps_step` must have the same shape."
+                )
+
+            if (self.eps_step > self.eps).any():
+                raise ValueError("The iteration step `eps_step` has to be smaller than the total attack `eps`.")
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:224]: I think this might only be true if `norm = np.inf` as in the removed check in line 212.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 4]
```diff
@@ -152,37 +162,15 @@ def _set_targets(self, x: np.ndarray, y: np.ndarray, classifier_mixin: bool = Tr
...
-        return mask
-
     def _check_params(self) -> None:
         super(ProjectedGradientDescentCommon, self)._check_params()
 
-        if self.eps_step > self.eps:
-            raise ValueError("The iteration step `eps_step` has to be smaller than the total attack `eps`.")
+        if isinstance(self.eps, float):
+            if self.eps_step > self.eps:
+                raise ValueError("The iteration step `eps_step` has to be smaller than the total attack `eps`.")
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:170]: This might only be true for `norm=np.inf`.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 5]
```diff
@@ -152,37 +162,15 @@ def _set_targets(self, x: np.ndarray, y: np.ndarray, classifier_mixin: bool = Tr
...
+        else:
+            if (self.eps_step > self.eps).any():
+                raise ValueError("The iteration step `eps_step` has to be smaller than the total attack `eps`.")
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:173]: This might only be true for norm=np.inf.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 6]
```diff
@@ -152,37 +162,15 @@ def _set_targets(self, x: np.ndarray, y: np.ndarray, classifier_mixin: bool = Tr
...
-                raise ValueError("Mask is only supported for classification.")
-
-        return mask
-
     def _check_params(self) -> None:
         super(ProjectedGradientDescentCommon, self)._check_params()
 
-        if self.eps_step > self.eps:
-            raise ValueError("The iteration step `eps_step` has to be smaller than the total attack `eps`.")
+        if isinstance(self.eps, float):
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:168]: Can we combine all conditions into a single if condition?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 7]
```diff
@@ -259,6 +247,9 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
         :type mask: `np.ndarray`
         :return: An array holding the adversarial examples.
         """
+        # Ensure eps is broadcastable
+        self._check_eps(x=x)
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:251]:
```suggestion
        self._check_compatibility_input_and_eps(x=x)
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 8]
```diff
@@ -127,6 +127,9 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
         """
         import torch  # lgtm [py/repeated-import]
 
+        # Ensure eps is broadcastable
+        self._check_eps(x=x)
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:131]:
```suggestion
        self._check_compatibility_input_and_eps(x=x)
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 9]
```diff
@@ -126,6 +126,9 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
         """
         import tensorflow as tf  # lgtm [py/repeated-import]
 
+        # Ensure eps is broadcastable
+        self._check_eps(x=x)
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:130]:
```suggestion
        self._check_compatibility_input_and_eps(x=x)
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @minhitbk Thank you very much for your pull request! I have a few comments and change requests, could you please take a look? Could you please also run the Black formatter with line length 120 in all modified files?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py (hunk 3 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py]: solved.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py (hunk 4 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py]: solved.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py (hunk 5 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py]: solved.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py (hunk 6 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py]: solved.

**Comment by minhitbk** [comment]
> Hi @minhitbk Thank you very much for your pull request! I have a few comments and change requests, could you please take a look? Could you please also run the Black formatter with line length 120 in all modified files?

@beat-buesser , thanks for the reviews. I solved all your comments and also applied black format on all changed files.

File: art/attacks/evasion/fast_gradient.py [hunk 10]
```diff
@@ -102,6 +102,21 @@ def __init__(
...
+        :param x: An array with the original inputs.
+        """
+        if isinstance(self.eps, np.ndarray):
+            if isinstance(self.estimator, ClassifierMixin):
+                # Ensure the eps array is broadcastable
+                if len(self.eps.shape) > len(x.shape) or self.eps.shape != x.shape[-len(self.eps.shape) :]:
+                    raise ValueError("The `eps` shape must be broadcastable to input shape.")
+
+            else:
+                raise ValueError("The `eps` array is only supported for classification.")
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:118]: Why is the broadcastable `eps` only supported for classification?

File: art/utils.py [hunk 11]
```diff
@@ -298,49 +298,87 @@ def projection(values: np.ndarray, eps: float, norm_p: Union[int, float, str]) -
...
+    original_shape: Optional[Tuple] = None,
+) -> np.ndarray:
     """
     Generate randomly `m x n`-dimension points with radius `radius` and centered around 0.
 
     :param nb_points: Number of random data points.
     :param nb_dims: Dimensionality of the sphere.
     :param radius: Radius of the sphere.
     :param norm: Current support: 1, 2, np.inf, "inf".
+    :param original_shape: The original shape of the input batch.
```
Comment (beat-buesser) [inline art/utils.py:347]: Do we still need `nb_dims` if we introduce `original_shape`?

File: art/attacks/evasion/fast_gradient.py [hunk 12]
```diff
@@ -253,11 +314,31 @@ def _check_params(self) -> None:
         if self.norm not in [1, 2, np.inf, "inf"]:
             raise ValueError('Norm order must be either 1, 2, `np.inf` or "inf".')
 
-        if self.eps <= 0:
-            raise ValueError("The perturbation size `eps` has to be positive.")
+        if (not (isinstance(self.eps, (int, float)) and isinstance(self.eps_step, (int, float)))) and (
+            not (isinstance(self.eps, np.ndarray) and isinstance(self.eps_step, np.ndarray))
+        ):
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:319]: This condition should only be true if `minimal=True`.

File: art/attacks/evasion/fast_gradient.py [hunk 13]
```diff
@@ -102,6 +102,21 @@ def __init__(
...
+    def _check_compatibility_input_and_eps(self, x: np.ndarray):
+        """
+        Check the compatibility of the input with `eps` and `eps_step` which are of the same shape.
+
+        :param x: An array with the original inputs.
+        """
+        if isinstance(self.eps, np.ndarray):
+            if isinstance(self.estimator, ClassifierMixin):
+                # Ensure the eps array is broadcastable
+                if len(self.eps.shape) > len(x.shape) or self.eps.shape != x.shape[-len(self.eps.shape) :]:
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:114]: I'm not sure about the right-hand side of the `or` because it does not allow `eps` of shape `(1, 4)` for `x` of shape `(n, 4)` although Numpy broadcasting rules allow it.

Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py:114]: Fixed.

File: art/attacks/evasion/fast_gradient.py (hunk 10 above)
Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py:118]: Fixed.

File: art/attacks/evasion/fast_gradient.py (hunk 12 above)
Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py]: Fixed.

File: art/utils.py (hunk 11 above)
Comment (beat-buesser) [reply art/utils.py]: Fixed.

**Review by beat-buesser — APPROVED** [review]
